keps/sig-storage/20190621-namespace-transfer-api.md

```diff
@@ -0,0 +1,230 @@
+
+### Non-Goals
+
+* Actual implementation of transferring an object between namespaces
+
+## Proposal
+
+Introduce an external `NamespaceTransferRequest` object:
+
+```yaml
+apiversion: v1alpha1
+kind: NamespaceTransferRequest
+metadata:
+    name: pvc-transfer-request
+    namespace:  destination-namespace
+spec:
+    source:
+        name: source-namespace
+        name: source-pvc
+        kind: PersistentVolumeClaim
+
+```
+
+Introduce an external `NamespaceTransferApproval` object:
+
+```yaml
+apiversion: v1alpha1
+kind: NamespaceTransferApproval
+metadata:
+    name: pvc-transfer-approval
+    namespace: source-namespace
+spec: 
+    source:
+        name: source-pvc
+        kind: PersistentVolumeClaim
+    targetNamespace: destination-namespace
+
+```
+
+Introduce a parameter to objects (including the existing `kubernetes.pkg.apis.core.types.go.ObjectMeta` structure 
+to indicate if an object is transferable:
+
+```go
+type ObjectMeta struct {
+      ...
+      Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	  // Transferable indicates whether the object implements the namespace transfer API.  By default
+	  // the controller responsible for creation of the resource will set this appropriately based on
+	  // whether or not it implements the namespace transfer API.  A user may also explicitly set this
+	  // to `false` if they wish to prevent any transfer operations against the particular resource being
+	  // created regardless of the controllers capabilities.
+	  // +optional
+	  Transferable bool `json:"export,omitempty" protobuf:"varint,1,opt,name=transferable"`
+	  ...
+}
+
+### Example Implementation of the API
+
+We have a controller for each type of object we want to transfer (e.g. PVC, VolumeSnapshots, AppSnapshot, etc.) -- this way the logic for how transfer happens is custom per object type.
+
+Proposed logic for PVC transfer controller:
+
+The existing PV/PVC controller can be modified to include new pvc-transfer-controller logic. It will be responsible for transferring PVCs across namespaces. It will do the following:
+
+1. Monitor NamespaceTransferApproval, NamespaceTransferRequest, and PVC objects.
+2. Wait for a destination PVC to be created where PVC.DataSource points to a NamespaceTransferRequest (LocalObjectReference).
+3. Wait for that NamespaceTransferRequest and a matching NamespaceTransferApproval objects to exist. Matching means the request.source and approval.source match, the request.namespace and approval.targetNamespace match, and the request.source.kind is type PVC (this logic could be put in a library so different transfer controllers can all reuse it).
+4. Wait for source PVC to have no pods referencing it.
+5. Do additional validation (e.g. maybe StorageClass has a new AllowTransfer that must be set to true).
+6. Update the PVC object to indicate it is no longer "available" to use.
+7. Update NamespaceTransferRequest.status to indicate that transfer has started.
```

Comment:
What will the .status API look like exactly? e.g. should it have a "completed" condition/phase too? Or should the objects be deleted on completion?

Reply:
@wongma7 For the transfer Objects themselves?  I'd propose the end state is failed or succeeded; basically follow the way pods work.